firestore/query.go

```diff
@@ -1032,6 +1222,30 @@ func (it *DocumentIterator) Stop() {
 // GetAll returns all the documents remaining from the iterator.
 // It is not necessary to call Stop on the iterator after calling GetAll.
 func (it *DocumentIterator) GetAll() ([]*DocumentSnapshot, error) {
+	result, err := it.GetAllWithOptions()
+	if result != nil {
+		return result.DocumentSnapshots, err
+	}
+	return nil, err
+}
+
+// GetAllWithOptionsResult is the result of call to GetAllWithOptions method
+type GetAllWithOptionsResult struct {
+	DocumentSnapshots []*DocumentSnapshot
+
+	// Query explain metrics. This is only present when ExplainOptions is provided.
+	ExplainMetrics *ExplainMetrics
+}
+
+// GetAllWithOptions returns all the documents remaining from the iterator.
+//
+// The query is run with provided options. The provided non-nil options override any options provided earlier
+// e.g. below code will run the query with ExplainOptions{Analyze: true}
+// query := client.Collection("cities").RunOptions(ExplainOptions{Analyze: false}).Where("a", "=", "b")
+// result, err := query.GetAllWithOptions(ExplainOptions{Analyze: true})
+//
+// It is not necessary to call Stop on the iterator after calling GetAllWithOptions.
+func (it *DocumentIterator) GetAllWithOptions(opts ...RunOption) (*GetAllWithOptionsResult, error) {
```

Comment:
Is this new function required? It seems a little dangerous to give multiple places to set options like this. I could see that making maintenance/testing difficult in the future

Reply:
Please see: [go/cfdb-fs-go-query-profiling](http://goto.google.com/cfdb-fs-go-query-profiling) and [go/cfdb-ds-go-query-profiling](http://goto.google.com/cfdb-ds-go-query-profiling)

Reply:
Updated the design